src/map.c

```diff
@@ -3,752 +3,619 @@
+            pathp[1].node = rbtnode_get_right(pathp->node);
+        }
+    }
+    pathp->node = node;
+    /* A loop invariant we maintain is that all nodes with
+     * out-of-date summaries live in path[0], path[1], ..., *pathp.
+     * To maintain this, we have to summarize node, since we
+     * decrement pathp before the first iteration.
+     */
+    assert(!rbtnode_get_left(node));
+    assert(!rbtnode_get_right(node));
+    /* Unwind. */
+    for (pathp--; (uintptr_t) pathp >= (uintptr_t) path; pathp--) {
+        map_node_t *cnode = pathp->node;
+        if (pathp->cmp < 0) {
+            map_node_t *left = pathp[1].node;
+            rbtnode_set_left(cnode, left);
+            if (!rbtnode_get_red(left))
+                return;
+            map_node_t *leftleft = rbtnode_get_left(left);
+            if (leftleft && rbtnode_get_red(leftleft)) {
+                /* Fix up 4-node. */
+                map_node_t *tnode;
+                rbtnode_set_black(leftleft);
+                rbtnode_rotate_right(map_node_t, link, cnode, tnode);
+                cnode = tnode;
+            }
+        } else {
+            map_node_t *right = pathp[1].node;
+            rbtnode_set_right(cnode, right);
+            if (!rbtnode_get_red(right))
+                return;
+            map_node_t *left = rbtnode_get_left(cnode);
+            if (left && rbtnode_get_red(left)) {
+                /* Split 4-node. */
+                rbtnode_set_black(left);
+                rbtnode_set_black(right);
+                rbtnode_set_red(cnode);
+            } else {
+                /* Lean left. */
+                map_node_t *tnode;
+                bool tred = rbtnode_get_red(cnode);
+                rbtnode_rotate_left(map_node_t, link, cnode, tnode);
+                rbtnode_set_color(tnode, tred);
+                rbtnode_set_red(cnode);
+                cnode = tnode;
+            }
+        }
+        pathp->node = cnode;
+    }
+    /* Set root, and make it black. */
+    rbtree->root = path->node;
+    rbtnode_set_black(rbtree->root);
 }
 
-static void map_fix_colors(map_t obj, map_node_t *node)
+static void rb_remove(map_internal_t *rbtree, map_node_t *node)
 {
-    /* If root, set the color to black */
-    if (node == obj->head) {
-        rb_set_color(node, RB_BLACK);
-        return;
+    rb_path_entry_t path[RB_MAX_DEPTH];
+    rb_path_entry_t *pathp = NULL;
+    rb_path_entry_t *nodep = NULL;
+    /* Wind. */
+    path->node = rbtree->root;
+    for (pathp = path; pathp->node; pathp++) {
+        int cmp = pathp->cmp = (rbtree->cmp)(node, pathp->node);
+        if (cmp < 0) {
+            pathp[1].node = rbtnode_get_left(pathp->node);
+        } else {
+            pathp[1].node = rbtnode_get_right(pathp->node);
+            if (cmp == 0) {
+                /* Find node's successor, in preparation for swap. */
+                pathp->cmp = 1;
+                nodep = pathp;
+                for (pathp++; pathp->node; pathp++) {
```

Comment:
Hi @jserv,

I came up with a solution, but it might cause minor slowdown.

```
for (int i = 0; path[i].node; i++) {
        pathp = &path[i];
        int cmp = pathp->cmp = (rbtree->cmp)(node, pathp->node);
        if (cmp < 0) {
            path[i + 1].node = rbtnode_get_left(pathp->node);
        } else {
            path[i + 1].node = rbtnode_get_right(pathp->node);
            if (cmp == 0) {
                /* Find node's successor, in preparation for swap. */
                pathp->cmp = 1;
                nodep = pathp;
                for (int j = i + 1; path[j].node; j++) {
                    path[j].cmp = -1;
                    path[j + 1].node = rbtnode_get_left(path[j].node);
                }
                break;
            }
        }
    }
    ...
    ```

What do you think?

Reply:
Use switch-case since we do know the value of comparisons.

Reply:
Should I use switch case with goto to exit the for loop?

Reply:
> Should I use switch case with goto to exit the for loop?

Yes, it should be intentionally applied.